src/desktop/Window.cpp

```diff
@@ -856,21 +856,53 @@ bool CWindow::isTagged(const std::string& tag, bool strict) {
 }
 
 void CWindow::applyTag(const std::string& tag, bool dynamic) {
-    bool        isSet    = true;
+
     std::string tag_real = dynamic ? tag + "*" : tag;
 
-    if (tag.starts_with("-")) { // unset
-        isSet    = false;
+    bool        tagged  = isTagged(tag_real, true);
+    bool        changed = true;
+    bool        setTag  = true;
+
+    if (tag_real.starts_with("-")) { // unset
         tag_real = tag_real.substr(1);
-    } else if (tag.starts_with("+")) // set
+        changed  = tagged;
+        setTag   = false;
+    } else if (tag_real.starts_with("+")) { // set
         tag_real = tag_real.substr(1);
-    else // toggle if without prefix
-        isSet = !isTagged(tag_real, true);
+        changed  = !tagged;
+    } else // toggle if without prefix
+        setTag = !tagged;
 
-    if (isSet)
+    if (!changed)
+        return;
+
+    if (setTag)
         m_tags.emplace(tag_real);
     else
         m_tags.erase(tag_real);
+
+    // re run tag rules after update tag
+    queueScanForTagMatch();
+}
+
+void CWindow::queueScanForTagMatch() {
+    // TODO: assume single thread for window operation
+    if (!m_tagScanSource)
+        m_tagScanSource = wl_event_loop_add_idle(
+            g_pCompositor->m_sWLEventLoop,
```

Comment:
Here window operation is assumed to be run in single thread, if that's not TRUE, there'll need to be an extra lock or something better as the fix